### ISSUE TITLE
Error compiling 1.7.0-beta4

### ISSUE THREAD

**Issue by utrenkner** [h2o/h2o#718]
Today, I wanted to give beta4 a try (still running 1.6.2). But compilation on FreeBSD 10.2 ended with these errors:
```
[ 79%] Building C object CMakeFiles/libh2o-evloop.dir/lib/handler/access_log.c.o
/home/test/h2o-1.7.0-beta4/lib/handler/access_log.c:471:22: error: no member named 'remote_user' in 'struct st_h2o_req_t'
            if (req->remote_user.base == NULL)
                ~~~  ^
/home/test/h2o-1.7.0-beta4/lib/handler/access_log.c:473:26: error: no member named 'remote_user' in 'struct st_h2o_req_t'
            RESERVE(req->remote_user.len * 4);
                    ~~~  ^
/home/test/h2o-1.7.0-beta4/lib/handler/access_log.c:372:14: note: expanded from macro 'RESERVE'
        if ((capacity) + element->suffix.len > line_end - pos) {                                                                   \
             ^
/home/test/h2o-1.7.0-beta4/lib/handler/access_log.c:473:26: error: no member named 'remote_user' in 'struct st_h2o_req_t'
            RESERVE(req->remote_user.len * 4);
                    ~~~  ^
/home/test/h2o-1.7.0-beta4/lib/handler/access_log.c:374:66: note: expanded from macro 'RESERVE'
            line = expand_line_buf(line, line_end - line, off + (capacity) + element->suffix.len);                                 \
                                                                 ^
/home/test/h2o-1.7.0-beta4/lib/handler/access_log.c:474:50: error: no member named 'remote_user' in 'struct st_h2o_req_t'
            pos = append_unsafe_string(pos, req->remote_user.base, req->remote_user.len);
                                            ~~~  ^
4 errors generated.
*** Error code 1

Stop.
make[2]: stopped in /usr/home/test/h2o-1.7.0-beta4
```

Does anybody have suggestions for how to solve this?

**Comment by syohex** [comment]
How did you build ? I cannot reproduce on my FreeBSD 10.2.

And your error is a bit strange. `h2o_req_t.remote_user` is defined [here](https://github.com/h2o/h2o/blob/v1.7.0-beta4/include/h2o.h#L730) and there is not wrapped by OS specific `#ifdef`.

Could you retry to build ?


#### Build Log

- [Build log](https://gist.github.com/syohex/8b6aa043cda0e0d08310#file-freebsd-h2o-1-7-0-beta4-build-log)
- [CMakeCache](https://gist.github.com/syohex/8b6aa043cda0e0d08310#file-cmakecache-txt)

#### My environment

```
$ uname -a
FreeBSD h2o 10.2-RELEASE FreeBSD 10.2-RELEASE #0 r286666: Wed Aug 12 15:26:37 UTC 2015     root@releng1.nyi.freebsd.org:/usr/obj/usr/src/sys/GENERIC  amd64
$ clang --version
FreeBSD clang version 3.4.1 (tags/RELEASE_34/dot1-final 208032) 20140512
Target: x86_64-unknown-freebsd10.2
Thread model: posix
$ cmake --version
cmake version 3.4.1
```

**Comment by utrenkner** [comment]
I tried several more times - also re-downloading the sources, just to be on the safe side... but the result stays the same.

It worked without any problems with 1.6.0 and 1.6.2

[Build log](https://gist.github.com/utrenkner/ac0d22102e49f51495a6)
[CMakeCache](https://gist.github.com/utrenkner/f77887a2d39b6cf37eb0)


My environment
```
# uname -a
FreeBSD xxxxxxxxx 10.2-RELEASE-p9 FreeBSD 10.2-RELEASE-p9 #0: Thu Jan 14 01:32:46 UTC 2016     root@amd64-builder.daemonology.net:/usr/obj/usr/src/sys/GENERIC  amd64
# clang --version
FreeBSD clang version 3.4.1 (tags/RELEASE_34/dot1-final 208032) 20140512
Target: x86_64-unknown-freebsd10.2
Thread model: posix
# cmake --version
cmake version 3.4.1

```

**Comment by syohex** [comment]
Can you build other 1.7 beta version, beta3, beta2, beta1 ?

**Comment by utrenkner** [comment]
Just tried out several things...

beta2 shows the exact same error (therefore, I did not even test beta3)
beta1 stops already before that with an error in the mruby_handler. I then tried to build without mruby, but even then it would stop there.

I also went back and checked, that I can still compile 1.6.2 and the result is positive.

Thus, between 1.6.x and 1.7.0-beta2 must have changed something that triggers the error in lib/handler/access_log.c

**Comment by syohex** [comment]
Could you try `git bisect` for finding bad commit of build failure on your environment ?

```
% git clone https://github.com/h2o/h2o
% cd h2o
% git checkout v1.7.0-beta2
% git bisect start
% git bisect good v1.6.2
# Do 'git bisect bad' if you build success, do 'git bisect bad' if you build failre
```

See also
- https://git-scm.com/book/en/v2/Git-Tools-Debugging-with-Git#Binary-Search
- https://git-scm.com/docs/git-bisect

**Comment by utrenkner** [comment]
Thank you for walking me through git bisect! This is a fantastic feature...!

Here is the (first) result: `e2f6d9ebe7bcede4454eeeebc748673dbf83fddc is the first bad commit`
(Author: Kazuho Oku <kazuhooku@gmail.com>; Date: Wed Dec 23 16:48:29 2015 +0900; implement HTTP client; it works!)

This are the data from the build at that commit:
[make.log](https://gist.github.com/utrenkner/c537bb51874161f49067)
[CMakeCache](https://gist.github.com/utrenkner/b540b6afe908d089a907)

**Comment by syohex** [comment]
Hmm, https://github.com/h2o/h2o/commit/e2f6d9ebe7bcede4454eeeebc748673dbf83fddc did not change both `lib/handler/access_log.c` and header files which are included from `access_log.c`.

**Comment by kazuho** [comment]
@utrenkner Can you please try #722 to see it fixes your issue?

**Comment by utrenkner** [comment]
Thanks @kazuho.

You seem to have identified the cause correctly, but the #722 does not fix it for me (I can post CCache and build log, should you need it).

It seems like other information from the existing h2o installation are still leaking through to the build, leading to the errors. To test this, I cloned the production FreeBSD jail and deleted all the files from the previous 1.6.2 installation. After that, the build (with #722) went through without any problems.

Some other observations:
- In an earlier update from 1.6.0 to 1.6.2 I noticed that when I ran `./h2o --version` on the new build, it still showed "1.6.0". Only after I had deleted the old installation, did the new build show the correct version "1.6.2".

- I also noticed, that even WITHOUT the compile flag -DWITH_MRUBY=ON, h2o was build WITH mruby. I tested and found that I explicitly had to compile with -DWITH_MRUBY=OFF if I wanted h2o without mruby support. This may be intended behavior, but it is not what I would have expected.

**Comment by kazuho** [comment]
@utrenkner Sorry to hear that the issue still persists.

Can you please post the log of CMake as well as `make VERBOSE=1`.?  That will help.  Also I would appreciate it if you could run cmake in a fresh directory (i.e. with no information cached by cmake).

> * In an earlier update from 1.6.0 to 1.6.2 I noticed that when I ran ./h2o --version on the new build, it still showed "1.6.0". Only after I had deleted the old installation, did the new build show the correct version "1.6.2".

Version number is coded in `h2o/version.h` that gets installed.  So what you are seeing is another evidence that the build referred to the installed version of the file.

> * I also noticed, that even WITHOUT the compile flag -DWITH_MRUBY=ON, h2o was build WITH mruby. I tested and found that I explicitly had to compile with -DWITH_MRUBY=OFF if I wanted h2o without mruby support. This may be intended behavior, but it is not what I would have expected.

In 1.4, mruby support was introduced as experimental and was off by default.  From 1.5, it is turned on by default (ref: https://github.com/h2o/h2o/blob/v1.7.0-beta4/Changes#L114).

We will be introducing new features this way; first adding something experimentally that people could opt-in to, and then turn it on by default.

Can you describe what made you think that mruby support is off by default; depending on the cause, it might be better for us to consider on how we update Changes, etc.

**Comment by utrenkner** [comment]
Hi @kazuho, thank you very much for your help - I really appreciate it!

>Can you please post the log of CMake as well as make VERBOSE=1.? That will help. Also I would appreciate it if you could run cmake in a fresh directory (i.e. with no information cached by cmake).

I always started building with a fresh directory, copied from the cloned github sources (and earlier, for the releases, I used the download links provided with in the release notes).

To be 100% I now deleted the two folders and started from scratch - the resulting logs are linked below. Just to document - this is what I did:
```
$ pwd
/tmp
$ rm -rf test
$ rm -rf h2o
$ git clone https://github.com/h2o/h2o.git
Cloning into 'h2o'...
remote: Counting objects: 19997, done.
remote: Compressing objects: 100% (47/47), done.
remote: Total 19997 (delta 12), reused 0 (delta 0), pack-reused 19950
Receiving objects: 100% (19997/19997), 27.31 MiB | 5.25 MiB/s, done.
Resolving deltas: 100% (12746/12746), done.
Checking connectivity... done.
$ cp -Rp h2o test
$ cd test
$ git checkout 0a0cf6ef7a87003b11dd25e0176b8f9ffae82264
Note: checking out '0a0cf6ef7a87003b11dd25e0176b8f9ffae82264'.
(...)
HEAD is now at 0a0cf6e... Merge pull request #722 from h2o/kazuho/issues/718
$ cmake -DCMAKE_INSTALL_PREFIX=/usr/local > ../cmake.log
$ make VERBOSE=1 > ../make.log
```
[Cmake log](https://gist.github.com/utrenkner/6b4100d7a1c332c4cf0e)
[CmakeCache](https://gist.github.com/utrenkner/d029c4aaec8370729ffa)
[Make log](https://gist.github.com/utrenkner/9c2bb3656c218f1eb291)

> Can you describe what made you think that mruby support is off by default; depending on the cause, it might be better for us to consider on how we update Changes, etc.
I think, I just missed the information that it is now switched on by default - if mruby is found on the server. I first did not notice that mruby was an external software that needed to be installed before using the mruby handler. h2o complained that it could not use the mruby_handler and I started to look for the steps necessary to have it working. Thus - probably through Google - found several pages that indicated the flat -DWITH_RUBY=ON. Specifically, I followed the recommendations I found on [calomel.org](https://calomel.org/h2o.html).

I realise that the h2o website does not (anymore) say to do this. It is not bad or wrong, but I find it rather unusual that by default, h2o compiles with mruby if it detects its presence - in most other software this would have to be explicitly configured in a config file / compile flag.

**Comment by syohex** [comment]
I think that this issue happens when both old h2o and OpenSSL are installed at same directory(In this case, I suppose it is `/usr/local`, both pkg and ports install packages at there).

For example OPEN_SSL_INCLUDE_DIR is `/usr/local/include`. [This](https://github.com/h2o/h2o/blob/4e1dbff509b57828eb4da4b554526250e8539b75/CMakeLists.txt#L292), [this](https://github.com/h2o/h2o/blob/4e1dbff509b57828eb4da4b554526250e8539b75/CMakeLists.txt#L293), and [this](https://github.com/h2o/h2o/blob/4e1dbff509b57828eb4da4b554526250e8539b75/CMakeLists.txt#L362) prepend `/usr/local/include` to include paths. So C compiler lookups `/usr/local/include/` ahead of `include/` and it uses older h2o header files.

I suppose following patch which appends(not preprend) OpenSSL include path can fix this issue.

```patch
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 25cde41..8a70be5 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -289,8 +289,8 @@ SET_TARGET_PROPERTIES(libh2o-evloop PROPERTIES
     OUTPUT_NAME h2o-evloop
     COMPILE_FLAGS "-DH2O_USE_LIBUV=0")
 IF (OPENSSL_FOUND)
-    TARGET_INCLUDE_DIRECTORIES(libh2o BEFORE PUBLIC ${OPENSSL_INCLUDE_DIR})
-    TARGET_INCLUDE_DIRECTORIES(libh2o-evloop BEFORE PUBLIC ${OPENSSL_INCLUDE_DIR})
+    TARGET_INCLUDE_DIRECTORIES(libh2o PUBLIC ${OPENSSL_INCLUDE_DIR})
+    TARGET_INCLUDE_DIRECTORIES(libh2o-evloop PUBLIC ${OPENSSL_INCLUDE_DIR})
     TARGET_LINK_LIBRARIES(libh2o ${OPENSSL_LIBRARIES})
     TARGET_LINK_LIBRARIES(libh2o-evloop ${OPENSSL_LIBRARIES})
     IF (LIBUV_FOUND AND NOT WITHOUT_LIBS)
@@ -359,7 +359,7 @@ IF (WITH_BUNDLED_SSL)
     ADD_DEPENDENCIES(h2o bundled-ssl)
 ELSE (WITH_BUNDLED_SSL)
     IF (OPENSSL_FOUND)
-        TARGET_INCLUDE_DIRECTORIES(h2o BEFORE PUBLIC ${OPENSSL_INCLUDE_DIR})
+        TARGET_INCLUDE_DIRECTORIES(h2o PUBLIC ${OPENSSL_INCLUDE_DIR})
         TARGET_LINK_LIBRARIES(h2o ${OPENSSL_LIBRARIES})
     ENDIF (OPENSSL_FOUND)
 ENDIF (WITH_BUNDLED_SSL)
```

**Comment by utrenkner** [comment]
@syohex Many thanks!!! This really does the trick.

I started with a clean copy of the git sources -> cmake -> patch -> make... and encountered no errors! 

```
$ ./h2o --version
h2o version 1.7.0-beta4
OpenSSL: OpenSSL 1.0.2e 3 Dec 2015
mruby: YES
```

**Comment by syohex** [comment]
I think this issue is fixed by #736 and I confirmed build error is gone.

**Comment by kazuho** [comment]
Yeah!  Thank you for the fix.
